### ISSUE TITLE
Discussion: Evaluate geoserver-cloud integration

### ISSUE THREAD

**Issue by mwallschlaeger** [GeoNodeUserGroup-DE/geonode-k8s#171]
## Task Description

As geoserver is basically the only service not scalable in geonode-k8s, there are currently two approaches to make this happen.
One mentioned by Geosolutions, loadbalancing incomming traffic to multiple geoserver instances. The other one would be to use the geoserver-cloud deployment from camptocamp. As far as I understand this can not be integrated directly as it is missing at least  the geoserver geonode plugin, which uses geonode to authorize geoserver login attempts. It might be possible to bypass this plugin by configuring geoserver to use oauth2 to authorize users with geonode. Does anybody has experiences or did further evaluation on including ether loadbalancing geoserver or geoserver cloud into geonode or geonode-k8s? @AlexGacon @ridoo

## Additional Information

Any additional information or context that may be helpful in completing the task.

**Comment by AlexGacon** [comment]
I work on daily basis with geoserver cloud so I would be happy to have it in geonode-k8s but there are several obstacles to do so:

- as you mentionned, the geonode plugins are not integrated/ported to geoserver-cloud. Not sure that your idea for bypassing the geonode oauth2 will work either.
- Geoserver Cloud does not support GeoFence. We have developed GeoServer ACL as a replacement but the REST API is not exactly the same as GeoFence so it cannot be drop like this in replacement.

Some of our customers ask for quotation on developing this but no progress so far.

Regarding the use of a cluster of GeoServer, it is possible but you have to put a mechanism to synchronize the configuration between the different GeoServer instances, since you cannot share a configuration folder between them on K8s. One option could be to store the GeoServer configuration into database (there are several community extensions for that); in this case you only have to find how to tell all the instances to update their configuration cache after a change in GeoNode (generally it means deploying somethings like JMX or RabbitMQ).

**Comment by mwallschlaeger** [comment]
@giohappy does geosolution has an idea on how to scale geoserver on kubernetes. By overcomming geonode plugins or by loadbalancing geoserver?

**Comment by ridoo** [comment]
@mwallschlaeger no deeper experience in clustering/loadbalancing geoserver so far.

ATM, the "geonode plugin" for GeoServer is quite opaque to me, but I may have missed the right entry point to dive into. I do know https://github.com/GeoNode/geoserver-geonode-ext/, though, it just provides some binary `jar` files and an empty data directory.

I have not looked into ACL and geofence, yet, so I cannot forsee, if oauth2 and ACL offer an alternative to standard geofence which might enables geoserver-cloud as an alternative to standard geoserver ... However, it would be very interesting to test out what would be possible with ACL and oauth2.

**Comment by ridoo** [comment]
For the record: https://github.com/geoserver/geoserver-cloud/pull/475

**Comment by AlexGacon** [comment]
@ridoo some of the extensions are already available in GS Cloud (wps, css)

**Comment by ridoo** [comment]
@AlexGacon are we able to create a "supported" matrix?

| plugin | supported by gs-cloud | comment | 
| ----------- | ------------------------------- | -------------- |
| [authkey](https://github.com/geoserver/geoserver/tree/main/src/extension/authkey) ([docs](https://docs.geoserver.org/latest/en/user/extensions/authkey/index.html)) | :question:  | |
| [backup-restore](https://github.com/geoserver/geoserver/tree/main/src/community/backup-restore) | :question: | |
| [control-flow](https://github.com/geoserver/geoserver/tree/main/src/extension/control-flow) ([docs](https://docs.geoserver.org/latest/en/user/extensions/controlflow/index.html))| :question: | |
| [css](https://github.com/geoserver/geoserver/tree/main/src/extension/css) | :heavy_check_mark: | [as indicated](https://github.com/zalf-rdm/geonode-k8s/issues/171#issuecomment-2124940558) |
| [excel](https://github.com/geoserver/geoserver/tree/main/src/extension/excel) ([docs](https://docs.geoserver.org/latest/en/user/extensions/excel.html))| :question:  | |
| [gdal](https://github.com/geoserver/geoserver/tree/main/src/extension/gdal) | :question:  | |
| [geofence-server](https://github.com/geoserver/geoserver/tree/main/src/extension/geofence/geofence-server) ([docs](https://docs.geoserver.org/latest/en/user/extensions/geofence-server/index.html)) | :question:  | resolved by https://github.com/geoserver/geoserver-acl? |
| [geofence-wps](https://github.com/geoserver/geoserver/tree/main/src/extension/geofence/geofence-wps) ([docs](https://docs.geoserver.org/latest/en/user/extensions/geofence-wps/index.html)) | :question: | |
| [kmlppio](https://github.com/geoserver/geoserver/tree/main/src/extension/wps/wps-kml-ppio) | :question:  | |
| [importer](https://github.com/geoserver/geoserver/tree/main/src/extension/importer) ([docs](https://docs.geoserver.org/latest/en/user/extensions/importer/index.html)) | :question:  | |
| [monitor](https://github.com/geoserver/geoserver/tree/main/src/extension/monitor) ([docs](https://docs.geoserver.org/latest/en/user/extensions/monitoring/index.html))| :question: | |
| [oauth2-geonode](https://github.com/geoserver/geoserver/tree/main/src/community/security/oauth2-geonode) | :question:  | resolved by https://github.com/geoserver/geoserver-cloud/pull/475? |
| [oauth2-openid-connect ](https://github.com/geoserver/geoserver/tree/main/src/community/security/oauth2-openid-connect)| :question:  | resolved by https://github.com/geoserver/geoserver-cloud/pull/475? |
| [printing](https://github.com/geoserver/geoserver/tree/main/src/extension/printing) ([docs](https://docs.geoserver.org/latest/en/user/extensions/printing/index.html)) | :question:  | |
| [querylayer](https://github.com/geoserver/geoserver/tree/main/src/extension/querylayer) ([docs](https://docs.geoserver.org/latest/en/user/extensions/querylayer/index.html)) | :question:  | |
| [sldService](https://github.com/geoserver/geoserver/tree/main/src/extension/sldService) ([docs](https://docs.geoserver.org/latest/en/user/extensions/sldservice/index.html)) | :question: | |
| [wmts-multi-dimensional](https://github.com/geoserver/geoserver/blob/main/src/extension/wmts-multi-dimensional) ([docs](https://docs.geoserver.org/latest/en/user/extensions/wmts-multidimensional/index.html)) | :question:  | |
| [wps](https://github.com/geoserver/geoserver/tree/main/src/extension/wps) | :heavy_check_mark:  | [as indicated](https://github.com/zalf-rdm/geonode-k8s/issues/171#issuecomment-2124940558) |
| [wps-download](https://github.com/geoserver/geoserver/tree/main/src/extension/wps-download) ([docs](https://docs.geoserver.org/latest/en/user/extensions/wps-download/index.html)) | :question:  |  |
| [ysld](https://github.com/geoserver/geoserver/tree/main/src/extension/ysld) | :question:  | |

**Comment by AlexGacon** [comment]
To me monitor and control-flow are not mandatory but a nice to have.

For monitoring, GS Cloud have metrics endpoint you can use with Prometheus or so (metrics per end-point, not as precise as the monitoring extension).

For control-flow, you can easily add new pods if required. We will also work soon on autoscaling.

**Comment by AlexGacon** [comment]
@mwallschlaeger @ridoo we are looking for fundings to work on a GS Cloud integration. Is it something which can be arranged on your side?

**Comment by groldan** [comment]
Hi, here's a working demo docker compose for geoserver cloud integration https://github.com/groldan/geonode-geoserver-cloud-compose
Note this is to check the Oauth2 functionality only.

Re the required extensions:

* authkey: integrated and enabled by default
* backup-restore: pending. When tried to integrate it had some bugs I don't recall. May try to resurrect it.
* control-flow: it probably makes more sense to do something in the gateway.
* css: integrated.
* excel: pending. Difficulty: easy
* gdal: pending. Difficulty: medium/complex. 
* geofence/wps: replaced by ACL in gs-cloud. The ACL REST API is different than Geofence's. The python client can be generated with the openapi tools.
* kmlppio: pending. Difficulty: easy
* importer: integrated in the web-ui service, but pending on the rest-service. In web-ui it needs to be enabled with the [webui.importer.enabled](https://github.com/geoserver/geoserver-cloud-config/blob/ee14531fc2a4fafae9fe2786d8688b54fda0c74e/geoserver.yml#L273) config property (i.e. ConfigMap, System prop, env variable)
* monitor: As mentioned by Alex, we'd rather work on spring-boot friendlier integrations (probably opentelemetry). There's the prometheus export and several other spring boot actuator endpoints for metrics and other stuff.
* oauth2-geonode
* printing: not integrated. May deserve its own microservice?
* querylayer: pending. Difficulty: easy
* sldService: pending. Difficulty: easy
* wmts-multi-dimentional: pending. Difficulty: dunno
* wps: `wps-service`
* wps-download: : pending. Difficulty: easy/medium. May just be a matter of having a prescribed pv/mount-point for shared tmp dirs?
* ysld: pending. Difficulty: easy

**Comment by mwallschlaeger** [comment]
@AlexGacon hi sorry for my late reply, I was on holiday. As the `geonode-k8s` maintainer i would really like to see the integration of geoserver cloud into this project. But as a devops engineer at my current position, I cannot see further benefits for our team to run geoserver-cloud within our geonode installation, therefore i do not see any possibilities to arrange any kind of funding for this purpose.

**Comment by ridoo** [comment]
> Hi, here's a working demo docker compose for geoserver cloud integration https://github.com/groldan/geonode-geoserver-cloud-compose

@groldan thanks for the hint. Will try it out.

> we are looking for fundings to work on a GS Cloud integration. Is it something which can be arranged on your side?

@AlexGacon I fear, we cannot arrange funding from our side ATM.

**Comment by anthieni** [comment]
Good morning all, we are working on an evaluation and development, already advanced, of the integration of GeoNode with GS-Cloud ACL where we will be releasing the first version, today we have it functional but we are doing the last tests. 
We are doing this with our own funding but also looking at potential customers for this version and for our version of [GeoExpress Cloud](https://kan.com.ar/productos/#portal). 

@groldan  is more than useful and it is working perfectly for us. 💯 

We have already separated the Mapstore Client for better scaling and looking at adding the html that is still in the Django/ jinja templates. We are also considering updating the tool versions. 

I hope that next weeks we will have news to share with you.

**Comment by mwallschlaeger** [comment]
I'm looking forward for geonode supports geoserver3 to use scalability features of geoserver 3 instead of geoserver cloud. therefor i close this issue for now
